gen/src/JsonElecResponse.cxx

```diff
@@ -0,0 +1,94 @@
+#include "WireCellGen/JsonElecResponse.h"
+
+#include "WireCellUtil/Persist.h"
+#include "WireCellUtil/NamedFactory.h"
+#include "WireCellUtil/Response.h"
+#include "WireCellUtil/Units.h"
+
+#include <iostream>
+#include <algorithm>
+
+WIRECELL_FACTORY(JsonElecResponse, WireCell::Gen::JsonElecResponse, WireCell::IWaveform, WireCell::IConfigurable)
+
+using namespace std;
+using namespace WireCell;
+
+// Generate waveform given a binning and a tablulated function
+WireCell::Waveform::realseq_t generate(const WireCell::Binning& tbins,
+    const Waveform::realseq_t& times, const Waveform::realseq_t& amps)
+{
+    const int nsamples = tbins.nbins();
+    WireCell::Waveform::realseq_t ret(nsamples, 0.0);
+    for (int ind = 0; ind < nsamples; ++ind) {
+        const double time = tbins.center(ind);
+        if (time < times.front() or time > times.back())
+            ret.at(ind) = 0;
+        else { // find the element that is greater or equal
+            auto it = std::lower_bound(times.begin(), times.end(), time);
+            auto k = it - times.begin();
+            if (*it == time) ret.at(ind) = amps.at(k);
+            else { // interpolation
+                ret.at(ind) = amps.at(k-1) + (amps.at(k)-amps.at(k-1))/(times.at(k)-times.at(k-1))*(time - times.at(k-1));
+            }
+        }
+    }
+    return ret;
+}
+
+Gen::JsonElecResponse::JsonElecResponse(int nticks,  double t0, double postgain,
+                                        double tick, std::string filename)
+{
+    m_cfg["filename"] = filename;
+    m_cfg["postgain"] = postgain;
+    m_cfg["start"] = t0;
+    m_cfg["tick"] = tick;
+    m_cfg["nticks"] = nticks;
+}
+WireCell::Configuration Gen::JsonElecResponse::default_configuration() const { return m_cfg; }
+void Gen::JsonElecResponse::configure(const WireCell::Configuration& cfg)
+{
+    m_cfg = cfg;
+    auto filename = m_cfg["filename"].asString();
+    if (filename.empty()) {
+        THROW(ValueError() << errmsg{"must supply a JsonElecResponse filename"});
+    }
+
+    auto top = Persist::load(filename);
+    auto jtimes = top["times"];
+    auto jamps = top["amplitudes"];
+    if (jtimes.size() != jamps.size()) {
+        THROW(ValueError() << errmsg{"inconsistent dimensions in file " + filename});
+    }
+    const int nsamp = jtimes.size();
+    m_times.resize(nsamp);
+    m_amps.resize(nsamp);
+
+    if (jtimes.isNull()) {
+        THROW(ValueError() << errmsg{"no values given in file " + filename});
+    }
+    const double tick = waveform_period();
+    for (int i=0; i<nsamp; i++) {
+        m_times.at(i) = jtimes[i].asFloat() * units::us;
+        m_amps.at(i) = jamps[i].asFloat() * units::mV / units::fC;
+    }
```

Comment:
This use of units is not correct.  Data in JSON/Jsonnet form is expected to already be in WCT's system of units.

In general, in C++ code only multiply a unit to a literal number, not something already in a variable.

Reply:
I see. I will give it a try on the wire-cell-python package to handle json inputs and the units.

Reply:
Just a reminder, in Python we can have like `from wirecell import units` and then have code like `x = 20*units.cm`.

https://github.com/WireCell/wire-cell-python/blob/master/wirecell/units.py

A similar module is available in Jsonnet so it would even be possible to provide `JsonElecResponse` data as a `.jsonnet` file holding values with explicit units instead of a `.json.bz2` file with implicit units. (But, JSON is okay.)

https://github.com/WireCell/wire-cell-toolkit/blob/master/cfg/wirecell.jsonnet

Reply:
I have removed the units in the c++ code.